### ISSUE TITLE
Any plan to add URL parser?

### ISSUE THREAD

**Issue by tatsuhiro-t** [nodejs/llhttp#7]
Do you have any plan to add URL parser to llhttp?
http-parser kindly includes URL parser which is very convenient.
Currently llhttp lacks this functionality.  I can use URL parser from http-parser, but some macro names conflict, so I have to provide a header file to declare URL parser struct and functions only.

**Comment by indutny** [comment]
That should be most certainly possible. The best way to do it would probably be through separate compilation target. I'll look into it ASAP.

**Comment by indutny** [comment]
What is the API that your use case requires?

**Comment by tatsuhiro-t** [comment]
I use the following API functions to parse URLs:

```c
/* Initialize all http_parser_url members to 0 */
void http_parser_url_init(struct http_parser_url *u);

/* Parse a URL; return nonzero on failure */
int http_parser_parse_url(const char *buf, size_t buflen,
                          int is_connect,
                          struct http_parser_url *u);
```

**Comment by tatsuhiro-t** [comment]
I need all fields which http_parser_url has because original URL cannot be reconstructed from http_parser_url if there are any missing fields.

**Comment by derekargueta** [comment]
👋 Just wanted to chip in that this would be super helpful to get [Envoy](https://github.com/envoyproxy/envoy) migrated to llhttp from http_parser (https://github.com/envoyproxy/envoy/issues/5155). We have some [utility URL code](https://github.com/envoyproxy/envoy/blob/master/source/common/http/utility.cc#L36) that depends on it.

**Comment by huming2207** [comment]
Any updates for this issue?

This is the only problem for me to migrate `http_parser` my project to `llhttp` .

**Comment by indutny** [comment]
@huming2207 unfortunately there is no update on this. Despite originally being shipped together with http_parser, there is no specific need for URL parser to live in llhttp itself.

It might be interesting to try taking `llparse` and the necessary boilerplate from `llhttp` to produce something like `llurl` for parsing URLS.

**Comment by andrew-aladev** [comment]
Guys, despite your experience I don't think that this http parser will be widely used. It depends on "llparse" bicycle (that has already been abandoned) and llvm. It is not tested even on 1% of original http parser.

**Comment by knopp** [comment]
@andrew-aladev, IIRC llhttp is default parser in nodejs since node 12, and it does not depend on llvm.

**Comment by bsergean** [comment]
@indutny we might try to help here. One question thought as I see 2 paths forward.

1. Should we try to add support directly in llhttp ?
2. Should we make a standalone llurl ?

I found https://github.com/nodejs/llhttp/blob/master/src/llhttp/url.ts  / should we use this as a start ?
(I noticed it does not seem to support parsing port numbers, but everything else (path, query, scheme, etc...) looks there).
